students/rick_halsell/lesson03/mailroom.py

```diff
@@ -0,0 +1,181 @@
+        listdonorsfunc()
+    elif len(selection) == 0: # returns to main menu if input is blank
+        mainmenufunc()
+    elif selection in donors_dictionary.keys(): # calls function if the donor is in the dictionary
+        donorindictfunc(selection)
+    elif selection not in donors_dictionary.keys(): # calls function if the donor is NOT in the dictionary
+        donornotindictfunc(selection)
+    else: # return to main menu
+        mainmenufunc()
+    return
+
+# Funtion to provide Thank you note options
+def thankyoufunc():
+    global selection
+    #print('You selected \"Send a Thank You\"')
+    print ("""
+    ****** Thank You Note Options ***************
+    1. Type 'List' to see a full list of donors
+    2. Type in a donor's full name
+    3. Press Enter Key to return to the Main Menu
+    *********************************************
+    """)
+    fullnamefunc()
+    return
+
+# Function to Generate Thank you note printed to screen
+def thankyounotefunc(selection):
+    print (f"""
+    To: {selection}@email.com
+    From: mailer-auto@donations.com
+    Subject: Recent Donation
+
+    Dear {selection},
+
+    Thank you for your generous donation.
+
+    Sincerely,
+
+    The Donation Collection Staff
+
+    ***********************************
+    """)
+    time.sleep(5)
+    return
+
+# Function to exit the program
+def exitfunc():
+    print('Exiting Program!')
+    time.sleep(1)
+    sys.exit()
+    return
+
+# Restart or Exit Program Function
+def executeAgainfunc():
+    while True:
+        RunAgain = input('Would you like to execute this script again? (Y/N): ').lower()
+        if not (RunAgain in ('yes', 'y')):
+            exitfunc()
+        if (RunAgain in ('yes', 'y')):
+            mainmenufunc()
+    return
+
+def reportfunc():
+    # assigns variables for easy formatting
+    left_aligned = "Donor Name"
+    center = "Total Given"
+    mid_right_aligned = "Num Gifts"
+    right_aligned = "Average Gift"
+
+    # formatting and printing report header
+    print(f"{left_aligned:<15} | {center:^10} | {mid_right_aligned:>10} | {right_aligned:>13}")
+    print('-'*60)
+    for selection, value in donors_dictionary.items():
+        totaldonationsgiven = len(value) # gives the total number of donations in dictionary associated with donor
+        donationtotal = sum(value) # gives total sum of all donations
+        donationtotal = float(donationtotal) # turns sun into a float
+        avgdonation =  round(donationtotal / totaldonationsgiven, 2) # round answer to the second decimal place
+        avgdonation = (f"{avgdonation:.2f}") # format string to print 2 decimal places
+        donationtotal = (f"{donationtotal:.2f}") # format string to print 2 decimal places
+        print(f"{selection:<15}{donationtotal:^24}{totaldonationsgiven:>3}{avgdonation:>18}") # format string for report
+        time.sleep(0.5) # slow down terminal printing for user
+    mainmenufunc() # back to main menu
+    return
```

Comment:
Only thing missing here is donors should be listed in descending order by total donations